cpp/src/arrow/util/bit_util.h

```diff
@@ -827,6 +834,164 @@ Status InvertBitmap(MemoryPool* pool, const uint8_t* bitmap, int64_t offset,
+  static void Visit(const Bitmap (&bitmaps)[N], Visitor&& visitor) {
+    VisitImpl(bitmaps, VisitedIs<internal::call_traits::argument_type<0, Visitor&&>>{},
+              std::forward<Visitor>(visitor));
+  }
+
+  const std::shared_ptr<Buffer>& buffer() const { return buffer_; }
+
+  /// offset of first bit relative to buffer().data()
+  int64_t offset() const { return offset_; }
+
+  /// number of bits in this Bitmap
+  int64_t length() const { return length_; }
+
+  /// string_view of all bytes which contain any bit in this Bitmap
+  util::basic_string_view<uint8_t> bytes() const {
+    auto byte_offset = offset_ / 8;
+    auto byte_count = BitUtil::CeilDiv(offset_ + length_, 8) - byte_offset;
+    return util::basic_string_view<uint8_t>(buffer_->data() + byte_offset, byte_count);
+  }
+
+  template <typename Word>
+  bool word_aligned() const {
+    auto words = this->words<Word>();
+    auto valid_addr = AsInt(buffer_->data());
+    auto words_addr = AsInt(words.data());
+    return words_addr >= valid_addr &&
+           words_addr + words.size() * sizeof(Word) <= valid_addr + buffer_->capacity();
+  }
+
+  /// string_view of all Words which contain any bit in this Bitmap
+  template <typename Word>
+  util::basic_string_view<Word> words() const {
+    auto bytes_addr = AsInt(bytes().data());
+    auto words_addr = bytes_addr - bytes_addr % sizeof(Word);
+    auto word_byte_count =
+        BitUtil::RoundUpToPowerOf2(bytes_addr + bytes().size(), sizeof(Word)) -
+        words_addr;
+    return util::basic_string_view<Word>(reinterpret_cast<const Word*>(words_addr),
+                                         word_byte_count / sizeof(Word));
+  }
+
+  /// offset of first bit relative to words<Word>().data()
+  template <typename Word>
+  int64_t word_offset() const {
+    return offset_ + (AsInt(buffer_->data()) - AsInt(words<Word>().data())) * 8;
+  }
+
+ private:
+  template <typename>
+  struct VisitedIs {};
+
+  static size_t AsInt(const void* ptr) {
+    return reinterpret_cast<size_t>(reinterpret_cast<const char*>(ptr));
+  }
+
+  template <size_t N, typename Visitor>
+  static void VisitImpl(const Bitmap (&bitmaps)[N], VisitedIs<std::bitset<N>>,
+                        Visitor&& visitor) {
+    auto bit_length = bitmaps[0].length();
+    std::bitset<N> bits;
+    for (int64_t bit_i = 0; bit_i < bit_length; ++bit_i) {
+      for (size_t i = 0; i < N; ++i) {
+        bits[i] = bitmaps[i].GetBit(bit_i);
+      }
+      visitor(bits);
+    }
+  }
+
+  template <typename Word, size_t N, typename Visitor>
+  static void VisitImpl(const Bitmap (&bitmaps)[N], VisitedIs<std::array<Word, N>>,
+                        Visitor&& visitor) {
+    constexpr size_t kBitWidth = sizeof(Word) * 8;
+
+    util::basic_string_view<Word> words[N];
+    int64_t offsets[N];
+    for (size_t i = 0; i < N; ++i) {
+      words[i] = bitmaps[i].template words<Word>();
+      offsets[i] = bitmaps[i].template word_offset<Word>();
```

Comment:
This is (necessarily; templates) a header file, do you mean with `<cassert>`'s assert? If I understand `word_offset` correctly, it would be impossible for it to return something outside that range. I could write a unit test which checks this

Reply:
Yes, I mean plain `assert`.

Reply:
I'll add the assert